enterprise/micronaut/src/org/netbeans/modules/micronaut/symbol/MicronautSymbolSearcher.java

```diff
@@ -0,0 +1,259 @@
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import javax.swing.Icon;
+import org.netbeans.api.java.project.JavaProjectConstants;
+import org.netbeans.api.project.FileOwnerQuery;
+import org.netbeans.api.project.Project;
+import org.netbeans.api.project.ProjectUtils;
+import org.netbeans.api.project.SourceGroup;
+import org.netbeans.modules.csl.api.IndexSearcher;
+import org.netbeans.modules.csl.api.Modifier;
+import org.netbeans.modules.csl.api.OffsetRange;
+import org.netbeans.modules.csl.spi.GsfUtilities;
+import org.netbeans.modules.csl.spi.ParserResult;
+import org.netbeans.modules.parsing.impl.indexing.CacheFolder;
+import org.netbeans.modules.parsing.spi.indexing.support.QuerySupport.Kind;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.FileUtil;
+import org.openide.filesystems.URLMapper;
+import org.openide.util.Exceptions;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ *
+ * @author Dusan Balek
+ */
+@ServiceProvider(service=IndexSearcher.class)
+public class MicronautSymbolSearcher implements IndexSearcher {
+
+    @Override
+    public Set<? extends Descriptor> getTypes(Project project, String textForQuery, Kind searchType, Helper helper) {
+        return Collections.emptySet();
+    }
+
+    @Override
+    public Set<? extends Descriptor> getSymbols(Project project, String textForQuery, Kind searchType, Helper helper) {
+        if (project == null || !textForQuery.startsWith("@")) {
+            return Collections.emptySet();
+        }
+        Set<Descriptor> symbols = new HashSet<>();
+        for (SourceGroup sg : ProjectUtils.getSources(project).getSourceGroups(JavaProjectConstants.SOURCES_TYPE_JAVA)) {
+            try {
+                FileObject cacheRoot = getCacheRoot(sg.getRootFolder().toURL());
+                if (cacheRoot != null) {
+                    Enumeration<? extends FileObject> children = cacheRoot.getChildren(true);
+                    while (children.hasMoreElements()) {
+                        FileObject child = children.nextElement();
+                        if (child.hasExt("mn")) { //NOI18N
+                            loadSymbols(child, symbols);
+                        }
+                    }
+                }
+            } catch (IOException ex) {}
+        }
+        return symbols;
+    }
+
+    private static void loadSymbols(FileObject input, Set<Descriptor> symbols) {
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(input.getInputStream(), StandardCharsets.UTF_8))) {
+            FileObject fo = null;
+            String line;
+            while ((line = br.readLine()) != null) {
+                if (line.startsWith("url: ")) { //NOI18N
+                    String url = line.substring(5);
+                    fo = URLMapper.findFileObject(URI.create(url).toURL());
+                    if (fo == null) {
```

Comment:
this seems to cause a stale index (file removed, index not updated) to be skipped, the entire source root is skipped. OK ?

Reply:
No - index is stored in `*.mn` files created per java source file. In case the source file is removed, only symbols for that particular file are skipped.